Index: main.c
===================================================================
--- main.c	(revision 55)
+++ main.c	(working copy)
@@ -898,6 +898,7 @@
         xhash_free(conn->routes);
 
         if(conn->key != NULL) free(conn->key);
+        if(conn->dkey != NULL) free(conn->dkey);
         free(conn);
     }
 
@@ -906,6 +907,8 @@
     if(xhash_iter_first(s2s->outq))
         do {
              xhash_iter_get(s2s->outq, NULL, xhv.val);
+             while (jqueue_size(q) > 0)
+                 out_pkt_free((pkt_t) jqueue_pull(q));
              free(q->key);
              jqueue_free(q);
         } while(xhash_iter_next(s2s->outq));
Index: out.c
===================================================================
--- out.c	(revision 55)
+++ out.c	(working copy)
@@ -564,6 +564,14 @@
     return 0;
 }
 
+void out_pkt_free(pkt_t pkt)
+{
+    nad_free(pkt->nad);
+    jid_free(pkt->from);
+    jid_free(pkt->to);
+    free(pkt);
+}
+
 /** send a packet out */
 int out_packet(s2s_t s2s, pkt_t pkt) {
     char *rkey;
Index: s2s.h
===================================================================
--- s2s.h	(revision 55)
+++ s2s.h	(working copy)
@@ -377,3 +377,5 @@
   dnscache_t *dns_val;
   dnsres_t *dnsres_val;
 };
+
+void out_pkt_free(pkt_t pkt);
