Index: mod_help.c
===================================================================
--- mod_help.c	(revision 43)
+++ mod_help.c	(working copy)
@@ -92,6 +92,20 @@
 
     for(jid = all; jid != NULL; jid = jid->next)
     {
+        if (jid_compare_full(pkt->from, jid) == 0) {
+            /* make a copy of the nad so it can be dumped to a string */
+            nad_t copy = nad_copy(pkt->nad);
+            if (!copy) {
+                log_write(mod->mm->sm->log, LOG_ERR, "%s:%d help admin %s is messaging sm for help! packet dropped. (unable to print packet - out of memory?)", ZONE, jid_full(jid));
+                continue;
+            }
+            char * xml;
+            int len;
+            nad_print(copy, 0, &xml, &len);
+            log_write(mod->mm->sm->log, LOG_ERR, "%s:%d help admin %s is messaging sm for help! packet dropped: \"%.*s\"\n", ZONE, jid_full(jid), len, xml);
+            nad_free(copy);
+            continue;
+        }
         log_debug(ZONE, "resending to %s", jid_full(jid));
         pkt_router(pkt_dup(pkt, jid_full(jid), jid_user(pkt->to)));
     }
